Implementation of gossipsub v1.1 - Hardening

vyzo opened the pull request:
This PR implements the Hardening extensions as defined by [Gossipsub v1.1 Specification](https://github.com/libp2p/specs/blob/master/pubsub/gossipsub/gossipsub-v1.1.md)

# What is left to do
- [x] plug peer score into gossipsub algorithms
- [x] flood publishing option
- [x] adaptive gossip disemmination
- [x] (re)transmission limit for IWANT requests
- [x] fill in the peer score calculation details
- [x] Increase the code coverage with unit tests (#281 #287)
- [x] Testground testing

vyzo committed:
gomod: use go-libp2p-core@peer-records and go-libp2p-peerstore@certified-addrs

vyzo committed:
gomod: use go-libp2p-core@peer-records and go-libp2p-peerstore@certified-addrs

yusefnapora committed:
track changes to peer records in -core

yusefnapora committed:
update PR branch dependencies

vyzo committed:
hook the tracer for peer scoring

vyzo committed:
peer score scaffolding

vyzo committed:
hook peer score into control message handlers

vyzo committed:
filter peers with negative score in initial mesh on join

vyzo committed:
heartbeat doPX is a more complex decision

vyzo committed:
hook score into heartbeat maintenance

willscott commented on gossipsub.go:
```diff
@@ -180,8 +210,14 @@ func (gs *GossipSubRouter) HandleRPC(rpc *RPC) {
-	iwant := make(map[string]struct{})
+	// we ignore IHAVE gossip from any peer whose score is below the gossip threshold
+	score := gs.score.Score(p)
+	if score < gs.gossipThreshold {
```
From the protocol description, I was expecting some recording of metadata from lower-reputation peers, so that data can be learned from them if still needed as a fallback when the peers above the gossip threshold fail. Does that happen somewhere else?

vyzo replied:
The gossip threshold is generally negative, so that we allow for some gossip disemination from low scoring peers. If the score drops below that, the peer is effectively graylisted until his negative score decays (if it does that).
So what you are asking for is exactly what is happening.

vyzo committed:
remove peers with negative score from fanout

vyzo committed:
keep D peer when pruning for oversubscription

vyzo committed:
implement flood publishing option

vyzo committed:
correctly implement flood publishing

Only applies to messages that have been published by our own peer!

vyzo committed:
increase D_score to 4

vyzo committed:
backoff when we prune too, refuse to reGRAFT backed off peers

vyzo committed:
router.Publish doesn't need the source, it takes the whole message now

vyzo committed:
allow gossipsub to graylist peers based on score

vyzo committed:
select good peers for PX, tune the px related limits

vyzo committed:
adaptive gossip dissemination

vyzo committed:
track peer (re)transmissions in message cache

vyzo committed:
limit message retransmissions through IWANT requests

vyzo committed:
peer score parameters

vyzo committed:
explicit Start for peer scoring to inject the router

so that we can unit test without a router.

vyzo committed:
peer score calculation

vyzo committed:
peer score refresh: counter decay and mesh activation

vyzo committed:
fix peer IP tracking structure

vyzo committed:
add nil check in Score function

necessary for when the score is not specified in gossipsub router.

vyzo committed:
respec decay to zero parameter in counter decay

vyzo committed:
respect the time in mesh cap

vyzo committed:
message delivery tracking

vyzo committed:
minimize hash table accesses

vyzo committed:
mesh tracking

vyzo committed:
peer tracking

vyzo committed:
track message delivery time within the validation pipeline

vyzo committed:
comment cosmetics

vyzo committed:
simplify duplicate mesh delivery tracking

vyzo committed:
more logical code ordering

vyzo committed:
clean up the logic of mesh delivery window

vyzo committed:
sticky mesh failure penalty parameter

vyzo committed:
finetune sticky mesh failure penalty parameter

vyzo committed:
double check delivery to aboid duplicate counts

vyzo committed:
implement message delivery struct details

vyzo committed:
implement getIPs and refreshIPs

vyzo committed:
implement periodic maintenance

vyzo committed:
peer score parameter validation

vyzo commented:
Summoning @yusefnapora @raulk @Stebalien @Kubuxu @whyrusleeping 

This is ready for review (modulo unit tests).

yusefnapora commented on pubsub.go:
```diff
@@ -356,7 +361,11 @@ func WithDiscovery(d discovery.Discovery, opts ...DiscoverOpt) Option {
 	return func(p *PubSub) error {
-		p.tracer = &pubsubTracer{tracer: tracer, pid: p.host.ID(), msgID: p.msgID}
+		if tracer != nil {
+			p.tracer.tracer = tracer
```
I think `p.tracer` may be nil here - I get a nil-dereference when I pass a tracer into this branch, but it works if I pass nil

vyzo replied:
oops, wrong check. it should be `if p.tracer != nil`....

fixing.

vyzo replied:
fixed.

vyzo committed:
fix bug in WithEventTracer

wrong nil check.

whyrusleeping commented on pubsub.go:
```diff
@@ -775,6 +784,12 @@ func (p *PubSub) handleIncomingRPC(rpc *RPC) {
 
+	// ask the router to vet the peer before commiting any processing resources
+	if !p.rt.AcceptFrom(rpc.from) {
+		log.Warningf("received message from router graylisted peer %s. Dropping RPC", rpc.from)
```
should we do something if we get a bunch of these?

vyzo replied:
we could blacklist i guess, but then we'd have to count stuff.

whyrusleeping commented on gossipsub.go:
```diff
@@ -384,9 +536,18 @@ func (gs *GossipSubRouter) Publish(from peer.ID, msg *pb.Message) {
 			continue
 		}
 
+		if gs.floodPublish && from == gs.p.host.ID() {
```
so this is adding in the flood publish option, flooding IFF its from us.

vyzo replied:
yes, otherwise it would be floodsub :p

Stebalien replied:
Do messages ever expire? If not, can't an attacker:

1. Collect a set of messages from us.
2. Wait for us to "forget" we sent the messages.
3. Feed us our own messages back.

Causing us to repeatedly flood our own messages out to the network (assuming they've collected enough of them)?

Stebalien replied:
We should drop our own messages if we receive them from the network (although I'm still concerned about the general replay issue).

vyzo replied:
That's a good suggestion, we should never re-broadcast our own messages from the network.

vyzo replied:
So the simplest ways to implement is to check if `ReceivedFrom` also matches our own host ID.
If it doesn't and it claims to be from us, we can drop it.

Stebalien replied:
Do we have any general protection against re-sending the same valid messages over and over in a cycle?

vyzo replied:
There is the timecache, which marks it for 2 minutes (configuration variable).
Outside of the timecach window it's up to the application to detect stale/replayed messages at validation time.

Stebalien replied:
Ah, got it. That's probably as good as we can get for now.

vyzo replied:
I have added the `ReceivedFrom` check; if a message has source as ourselves and was not received from ourselves, it is ignored and not forwarded.

whyrusleeping commented on score.go:
```diff
@@ -0,0 +1,885 @@
+	go ps.background(gs.p.ctx)
+}
+
+func (ps *peerScore) Score(p peer.ID) float64 {
```
As far as I can tell, this gets called for every single message received. It seems to me like it may get rather resource intensive. Let's keep this in mind and observe CPU profiles during tests

vyzo replied:
Yes, the `Score` method gets called a lot. It should be relatively lightweight to compute, but good point about looking at CPU profiles.

Kubuxu replied:
We also should watch lock contention there.

schomatis commented on score.go:
```diff
@@ -0,0 +1,885 @@
+	}
+
+	// check that we have an app specific score; the weight can be anything (but expected positive)
+	if p.AppSpecificScore == nil {
```
Let's also check `AppSpecificWeight` here, or maybe even remove it altogether, as this is an app-controlled function that can regulate its own score values (in contrast with the other weights that modulate pubsub behavior).

vyzo replied:
That's a good suggestion, we can consider removing the app specific weight and just keep the score function.

mcdee replied:
+1 to removing `AppSpecificWeight`, I don't see it adding anything.

schomatis commented on gossipsub.go:
```diff
@@ -238,17 +351,44 @@ func (gs *GossipSubRouter) handleIWant(p peer.ID, ctl *pb.ControlMessage) []*pb.
+		}
+
+		// check the score
+		if score < 0 {
```
In general, from a security perspective, I would prefer if all the `score` validation code would happen at a single point, similar to what is being done in `AcceptFrom`. This is not a blocker for me because I understand that it's not as straightforward as that other case, here different handlers will react slightly differently to a blacklisted (or similar) peer.

Citing this particular line as an example of the rationale behind this comment (but this logic replicates in other places). If `score < 0` then we basically won't be doing any GRAFT or PX (seems we won't be doing much besides the prune control response). This is not at all evident buried here, and it makes it harder to audit and maintain, even if it's correct (which seems to be).

Ideally handlers like these one should deal with the positive logic, what actually gets done for accepted peers, and move all security checks upstream. Again, just a suggestion because there may be other repercussions for the proposed change I'm not seeing.

vyzo replied:
it's a matter of some code duplication and convenience, but we can certainly check upfront here.

vyzo replied:
ok, I tried it and the code gets convoluted and requires a goto. Not a very good tradeoff.

schomatis replied:
@vyzo Why would you need a goto? Could you push that code in another branch please? I think I may have not expressed this correctly and we may thinking of different refactorings.

schomatis commented on gossipsub.go:
```diff
@@ -70,6 +77,62 @@ func NewGossipSub(ctx context.Context, h host.Host, opts ...Option) (*PubSub, er
+//  implementing an effective graylist according to peer score.
+//
+// These thresholds should generally be negative, allowing some information to disseminate from low
+//  scoring peers.
```
Let's add checks for these and emit a warning please, e.g., it seems that almost always we would want a graylist threshold below 0 and any value above this range will likely be a programmer error.

schomatis replied:
From reading the spec it seems we should also check that gossip > publish > graylist (especially because this relation doesn't evident from reading the code).

vyzo replied:
Yeah, I'll add a check in validate.

vyzo replied:
done.

schomatis commented on gossipsub.go:
```diff
@@ -103,9 +182,19 @@ func (gs *GossipSubRouter) Protocols() []protocol.ID {
 	gs.tracer = p.tracer
+
+	// start the scoring, if any
+	if gs.score != nil {
```
Could we invert this and turn it on by default with some off-the-shelf reasonable values and make the consumer explicitly disable it if needed?

vyzo replied:
I can just move the nil check inside the `Start` implementation so that it does not occlude the logic.

vyzo replied:
I am not sure there we can come up with reasonable off-the-shelf defaults, let me think about it.

vyzo replied:
for now I have moved this check inside `Start` so that it doesn't occlude the logic.

vyzo replied:
So I have given it some more thought and I think we can make a gossipsub v1.1 constructor function that can fill in default parameters given the expected message rate and the application score function.

schomatis commented on gossipsub.go:
```diff
@@ -570,22 +746,47 @@ func (gs *GossipSubRouter) heartbeat() {
 				_, inMesh := peers[p]
 				_, doBackoff := backoff[p]
-				return !inMesh && !doBackoff
+				score := gs.score.Score(p)
```
Could we reuse the already computed scores in `scores` here?

vyzo replied:
no, we have scores for peers _in_ the mesh, not the other ones. But the code should short-circuit and avoid calling `Score` if `inMesh` or `doBackoff` is true.

vyzo replied:
I short-circuited the score calculation to only occur if `!inMesh && !doBackoff`.

schomatis commented on gossipsub.go:
```diff
@@ -630,10 +833,11 @@ func (gs *GossipSubRouter) heartbeat() {
 		for p := range peers {
 			_, ok := gs.p.topics[topic][p]
-			if !ok {
+			score := gs.score.Score(p)
```
Same.

vyzo replied:
again here we can short-circuit.

vyzo replied:
done.

yusefnapora commented on score.go:
```diff
@@ -0,0 +1,885 @@
+	drec.peers = nil
+}
+
+func (ps *peerScore) DuplicateMessage(msg *Message) {
```
@vyzo I'm getting some "assignment to nil map" errors running against this branch in testground, e.g.

```
   2.0120s      ERROR << c694df9af4d2 >> panic: assignment to entry in nil map
   2.0121s      ERROR << c694df9af4d2 >>
   2.0121s      ERROR << c694df9af4d2 >> goroutine 98 [running]:
   2.0121s      ERROR << c694df9af4d2 >> github.com/libp2p/go-libp2p-pubsub.(*peerScore).DuplicateMessage(0xc000091340, 0xc00025ebd0)
   2.0122s      ERROR << c694df9af4d2 >> 	/go/pkg/mod/github.com/libp2p/go-libp2p-pubsub@v0.2.7-0.20200311004457-925ed3894d0b/score.go:653 +0x14f
   2.0122s      ERROR << c694df9af4d2 >> github.com/libp2p/go-libp2p-pubsub.(*pubsubTracer).DuplicateMessage(0xc000091300, 0xc00025ebd0)
   2.0122s      ERROR << c694df9af4d2 >> 	/go/pkg/mod/github.com/libp2p/go-libp2p-pubsub@v0.2.7-0.20200311004457-925ed3894d0b/trace.go:106 +0x3a1
   2.0122s      ERROR << c694df9af4d2 >> github.com/libp2p/go-libp2p-pubsub.(*validation).validate(0xc000389290, 0x0, 0x0, 0x0, 0xc0000a8f90, 0x26, 0xc00025ebd0)
   2.0123s      ERROR << c694df9af4d2 >> 	/go/pkg/mod/github.com/libp2p/go-libp2p-pubsub@v0.2.7-0.20200311004457-925ed3894d0b/validation.go:207 +0x6ad
   2.0133s      ERROR << c694df9af4d2 >> github.com/libp2p/go-libp2p-pubsub.(*validation).validateWorker(0xc000389290)
   2.0133s      ERROR << c694df9af4d2 >> 	/go/pkg/mod/github.com/libp2p/go-libp2p-pubsub@v0.2.7-0.20200311004457-925ed3894d0b/validation.go:184 +0x7c
   2.0133s      ERROR << c694df9af4d2 >> created by github.com/libp2p/go-libp2p-pubsub.(*validation).Start
   2.0134s      ERROR << c694df9af4d2 >> 	/go/pkg/mod/github.com/libp2p/go-libp2p-pubsub@v0.2.7-0.20200311004457-925ed3894d0b/validation.go:98 +0x70
   2.3247s INCOMPLETE << c694df9af4d2 >>
```

This must be because `drec.peers` was released in an earlier `RejectMessage` call. Can you add a nil check here for `drec.peers` and exit early if it's already been rejected? Or do we actually want to keep tracking duplicate deliveries after we've rejected the message?

vyzo replied:
hrm, we do want to keep track of rejected duplicates, but I am not sure this is the case here.
Are you generating any rejected messages?

vyzo replied:
It's in a valid delivery.

vyzo replied:
I think we might have some throttling going on which can also set the peers map to nil.

vyzo replied:
ie, it is possible to start validation and also throttle validation of a duplicate because it can't push to the validation workers.

vyzo replied:
yeah, that's what I think is happening: the validation queue throttles before we enter validation, which rejects the message (with a message ID of unknown validity) and sets the peer map to nil.
In the meantime there is another validation happening that completes successfully, and boom.
Anyhow, a nasty little bug.

vyzo replied:
it should be fixed in https://github.com/libp2p/go-libp2p-pubsub/pull/263/commits/5f1b070d5922dc076572cbd03010766a6f8b89cf

vyzo committed:
distinguish throttled validation because of a full worker queue.

vyzo commented:
also merged in #268 as this fixes a tracing bug.

vyzo committed:
short-circuit score computation in peer selection filters

vyzo committed:
move score starting nil check inside the score implementation

vyzo committed:
validate the threshold values

vyzo committed:
add option to periodically inspect peer scores for debugging purposes

vyzo commented:
Add a `WithPeerScoreDebug` option to enable periodic inspection/dumping of the peer scores.

vyzo committed:
shuffle IWANT list

vyzo committed:
refresh backoff when pruning during the backoff period

vyzo committed:
add backoff when pruning because of negative score

vyzo committed:
check exact expiration date of backoff when pruning because of backoff

vyzo committed:
track heartbeat ticks, only clear backoff every 15.

vyzo commented:
Some improvements of the backoff handling:
- we now extend the backoff when pruning because of backoff or a negative score; we also check the exact expiration.
- backoff is cleaned up every 15 heartbeat ticks to amortize the cost of iterating over the maps.

Stebalien commented on gossipsub.go:
```diff
@@ -238,17 +368,49 @@ func (gs *GossipSubRouter) handleIWant(p peer.ID, ctl *pb.ControlMessage) []*pb.
+		if score < 0 {
+			// we don't GRAFT peers with negative score
+			log.Debugf("GRAFT: ignoring peer %s with negative score [score = %f, topic = %s]", p, score, topic)
+			// we do send them PRUNE however, because it's a matter of protocol correctness
```
FYI, this does leak information and immediately tells the attacker that they failed and should try again.

vyzo replied:
They will be backed off and if they try repeatedly immediately after they will never get in because they will keep extending their backoff period.
Otherwise they will have to wait at least 1min to try again.

I think we should still emit the `PRUNE` message here, because it's a matter of protocol correctness.

Consider an honest node that happens to have a negative score (eg because of staked participation and them not having put in the stake yet, or because it was in the mesh and got pruned because it was not delivering enough messages). If we don't emit a PRUNE here they will think they are in the mesh while they are not and fail to build a mesh.

Stebalien commented on gossipsub.go:
```diff
@@ -434,12 +619,20 @@ func (gs *GossipSubRouter) Join(topic string) {
+			if gs.score.Score(p) < 0 {
+				delete(gmap, p)
+			}
+		}
```
Why do this _here_? To avoid having to eagerly maintain `gs.fanout`?

vyzo replied:
yes indeed.

Stebalien commented on score.go:
```diff
@@ -0,0 +1,952 @@
+	// Note: In order to simulate many IPs in a managable manner when testing, you can set the weight to 0
+	//       thus disabling the IP colocation penalty.
+	IPColocationFactorWeight    float64
+	IPColocationFactorThreshold int
```
Note: this doesn't work well for IPv6 at all so I'm not sure how useful it is in general. I assume an attacker would just disable IPv4 and pick IPv6 addresses out of their ISP assigned address range at will. ISPs are supposed to give out 64bit ranges.

vyzo replied:
We can use `/64`s for IPv6 addresses, that's an easy fix.

vyzo replied:
In general it is _very_ useful as it makes it very hard to launch a sybil attack using just a few IP addresses (which is something that has been observed in the filecoin testnet).

vyzo replied:
I think we need to track both the IP and the /64 in the case of IPv6.
We can also add a whitelist for eg datacenters, so that those /64 don't get penalized.

Stebalien replied:
I wonder if a better approach would be to simply say that scores "bleed" between peers with similar IP addresses. That would allow us to say things like "everyone in this IP block appears to be bad, let's reduce their collective score".

Of course, that could be an attack vector...

vyzo replied:
Implemented the `/64` and whitelisting for IPv6.

Stebalien commented on score.go:
```diff
@@ -0,0 +1,952 @@
+			continue
+		}
+
+		ip6, err := remote.ValueForProtocol(ma.P_IP6)
```
Note: this function is dangerous as it'll pick out the first IP address, regardless of what comes first. I can't think of a case where it'll fail at the moment, but it's best to avoid it and call something like `manet.ToIP`.

vyzo replied:
ok, can do.

vyzo replied:
done.

Stebalien commented on score.go:
```diff
@@ -0,0 +1,952 @@
+	ps.Lock()
+	defer ps.Unlock()
+
+	// TODO: the reasons should become named strings; good enough for now.
```
Why not just fix it now?

vyzo replied:
sure, I can fix it now.

vyzo replied:
done.

Stebalien commented on score.go:
```diff
@@ -0,0 +1,952 @@
+	if reason == "validation throttled" {
+		// if we reject with "validation throttled" we don't penalize the peer(s) that forward it
+		// because we don't know if it was valid.
+		drec.status = delivery_throttled
```
nit: non-idiomatic go variable name.

vyzo replied:
it's an unexported constant.

Stebalien replied:
https://github.com/golang/go/wiki/CodeReviewComments#mixed-caps

vyzo replied:
sigh... done.

raulk commented on gossipsub.go:
```diff
@@ -70,6 +77,74 @@ func NewGossipSub(ctx context.Context, h host.Host, opts ...Option) (*PubSub, er
+//
+// These thresholds should generally be negative, allowing some information to disseminate from low
+//  scoring peers.
+func WithPeerScore(params *PeerScoreParams, gossipThreshold, publishThreshold, graylistThreshold float64) Option {
```
Can we encapsulate the thresholds within a struct too? That would make things more extensible in the future.

BTW – I like the separation between params that drive peer score calculation, and params that drive compensatory actions.

vyzo replied:
Sure, we can do that.

vyzo replied:
done.

raulk commented on score.go:
```diff
@@ -0,0 +1,952 @@
+	invalidMessageDeliveries float64
+}
+
+type peerScore struct {
```
Godocs here would be much appreciated 🙏 Describe the role of this struct, when/where it's used, and it's locking requirements (contention?), would be great.

raulk commented on score.go:
```diff
@@ -0,0 +1,952 @@
+		inspectScores = ticker.C
+	} else {
+		// never fires
+		inspectScores = make(chan time.Time)
```
nil would work as well.

vyzo replied:
ok.

vyzo replied:
done.

raulk commented on score.go:
```diff
@@ -0,0 +1,952 @@
+	defer gcDeliveryRecords.Stop()
+
+	var inspectScores <-chan time.Time
+	if ps.inspect != nil {
```
I would probably be useful to `defer` an inspection, so that we dump the scores one last time when the background process ends.

vyzo replied:
we can do that.

vyzo replied:
done.

raulk commented on score.go:
```diff
@@ -0,0 +1,952 @@
+	// P1: time in the mesh
+	// This is the time the peer has ben grafted in the mesh.
+	// The value of of the parameter is the time/TimeInMeshQuantum, capped by TimeInMeshCap
+	// The weight of the parameter MUST be positive.
```
or zero.

vyzo replied:
fixed.

raulk commented on score.go:
```diff
@@ -0,0 +1,952 @@
+	// This is the number of message deliveries in the topic.
+	// The value of the parameter is a counter, decaying with FirstMessageDeliveriesDecay, and capped
+	// by FirstMessageDeliveriesCap.
+	// The weight of the parameter MUST be positive.
```
or zero.

vyzo replied:
fixed.

raulk commented on score.go:
```diff
@@ -0,0 +1,952 @@
+	// P4: invalid messages
+	// This is the number of invalid messages in the topic.
+	// The value of the parameter is a counter, decaying with InvalidMessageDeliveriesDecay.
+	// The weight of the parameter MUST be negative.
```
or zero.

vyzo replied:
fixed.

raulk commented on score.go:
```diff
@@ -0,0 +1,952 @@
+	TopicWeight float64
+
+	// P1: time in the mesh
+	// This is the time the peer has ben grafted in the mesh.
```
been.

raulk commented on score.go:
```diff
@@ -0,0 +1,952 @@
+	TimeInMeshCap     float64
+
+	// P2: first message deliveries
+	// This is the number of message deliveries in the topic.
```
Can we specify why the "first" adjective?

raulk commented on score.go:
```diff
@@ -0,0 +1,952 @@
+
+	// P2: first message deliveries
+	// This is the number of message deliveries in the topic.
+	// The value of the parameter is a counter, decaying with FirstMessageDeliveriesDecay, and capped
```
FirstMessageDeliveriesDecay seems like a decay factor. Can we specify what the decay function is?

vyzo replied:
Huh? Decay is multiplicative for all counters.

raulk replied:
Then let's clarify that! ;-) These *Decay fields are factors that could be applied linearly, polynomially, exponentially, the reader doesn't know with

raulk commented on score.go:
```diff
@@ -0,0 +1,952 @@
+	// when validation succeeds.
+	// This window accounts for the minimum time before a hostile mesh peer trying to game the score
+	// could replay back a valid message we just sent them.
+	// It effectively tracks first and near-first deliveries, ie a message seen from a mesh peer
```
Expand on "near-first"?

raulk commented on score.go:
```diff
@@ -0,0 +1,952 @@
+
+type TopicScoreParams struct {
+	// The weight of the topic.
+	TopicWeight float64
```
Does the sum of weights have to be 1?

vyzo replied:
No, there is no requirement to normalize the weight, so it is up to the application.

raulk commented on score.go:
```diff
@@ -0,0 +1,952 @@
+	ma "github.com/multiformats/go-multiaddr"
+)
+
+type PeerScoreParams struct {
```
Can we move the two Params structs to a `score_params.go` file?

vyzo replied:
ok...

vyzo replied:
done.

raulk commented on score.go:
```diff
@@ -0,0 +1,952 @@
+// When this option is enabled, the supplied function will be invoked periodically to allow
+// the application to inspec or dump the scores for connected peers.
+// This option must be passed _after_ the WithPeerScore option.
+func WithPeerScoreDebug(inspect PeerScoreInspectFn, period time.Duration) Option {
```
nit: homogeneous naming. `WithPeerScoreInspect`? Or `PeerScoreDumpFn`/`WithPeerScoreDump`?

vyzo replied:
renamed to `WithPeerScoreInspect`.

raulk commented on score.go:
```diff
@@ -0,0 +1,952 @@
+	for _, params := range p.Topics {
+		err := params.validate()
+		if err != nil {
+			return err
```
Can we wrap this error to annotate with the topic name that failed validation?

vyzo replied:
Sure.

vyzo replied:
done.

vyzo committed:
fix rebase artifacts

vyzo committed:
gomod tidy

vyzo commented:
rebased and regargeted on master as #234 has merged.

raulk commented on score.go:
```diff
@@ -0,0 +1,952 @@
+	return ps.score(p)
+}
+
+func (ps *peerScore) score(p peer.ID) float64 {
```
Eventually, we might find that we'd benefit from caching the score, and recalculating it in batch every N ticks (resolution), under a global lock. Recalculating the score on access may be heavy and may introduce minor skews that may create subtle unfairness across peers based on access time.

vyzo replied:
I thought about it, but I think it's premature to do it now.

raulk commented on score.go:
```diff
@@ -0,0 +1,952 @@
+
+		// P1: time in Mesh
+		if tstats.inMesh {
+			p1 := float64(tstats.meshTime / topicParams.TimeInMeshQuantum)
```
We should globally cap the sum total as well. Else, an attacker could boost their score artificially by meshing with us on many topics, and staying on those for a long time.

vyzo replied:
We already have an implicit global cap: each individual topic score is capped, so the global cap is the weight sum of the caps.

raulk replied:
I was thinking of a scenario where an attacker could enumerate all topics we're interested in, and GRAFT with us on all of them, whereas the average honest peer only GRAFTs with us on a subset of topics. That attacker could accumulate a higher score by the mere fact that they're behaving well on more topics than the rest. But I guess that's how the algorithm is designed to work? Also, the probability of us accepting all of their `GRAFT`s would be low.

vyzo replied:
On further thought it would make sense to cap the topic score contribution; I'll add an (optional) total topic score cap parameter.

vyzo replied:
Implemented an (optional) TopicScoreCap parameter that produces the desired effect.

raulk commented on score.go:
```diff
@@ -0,0 +1,952 @@
+		if peersInIP > ps.params.IPColocationFactorThreshold {
+			surpluss := float64(peersInIP - ps.params.IPColocationFactorThreshold)
+			p6 := surpluss * surpluss
+			score += p6 * ps.params.IPColocationFactorWeight
```
Comment: this weight is negative -- it is a penalty score.

vyzo replied:
the negative comment exists in the parameters struct, it's redundant here.

raulk replied:
I'm asking for a code comment, not API docs. I'll add it myself. (We'll all be maintaining this code in the future.)

raulk commented on score.go:
```diff
@@ -0,0 +1,952 @@
+				delete(ps.peerStats, p)
+			}
+
+			// we don't decay retained scores, as the peer is not active.
```
Can a peer disconnect and connect seconds before the retention period elapses, in order to keep their score "alive" perpetually? We may want to introduce additional measures here, e.g.:

1. Tallying periods of connectedness/disconnectedness across interruptions. When the total disconnectedness time surprasses a threshold, we forget the peer's score.
2. Tracking absolute number of disconnections and penalising for interruptions, relative to interruptions from other peers => rationale: if we have a high interruption rate from all peers, we might be the ones that are unstable and don't want to penalise anybody because we'll be shooting ourselves in the foot.

vyzo replied:
We could, but that just adds extra complexity here. Can we punt on this until we have a clear idea that it is a problem and how to resolve it?

raulk replied:
Complexity how? Counting connections and disconnections is simple; aren't we already subscribing to connectedness events, anyway?

Anyway, I think it's trivial to game the scoring system right now, and it shouldn't be this simple. cc @yiannisbot @daviddias

vyzo replied:
That's a heavy statement. How would you game the score by connecting and disconnecting?

vyzo replied:
we can make it so that only negative scores are retained if you are so concerned about this.

raulk replied:
> How would you game the score by connecting and disconnecting?

1. Attacker gets GRAFT'ed onto a topic.
2. Attacker behaves well, accumulates positive score.
3. Attacker disconnects. Their positive score is retained for `RetainScore` time.
4. Just before `RetainScore` elapses, the attacker connects again. This refreshes their expiry time.
5. They wait a little, so that a score recalculation kicks in. This line will actually award them points for the time they've been disconnected: https://github.com/libp2p/go-libp2p-pubsub/blob/feat/hardening/score.go#L517 (EDIT: or maybe not? I don't know if we reset P1 on an interruption).
6. They disconnect.
7. Lather, rinse, repeat. You keep building up your P1 score without even being connected.

vyzo replied:
Ok, that's a valid concern. The simple (and correct) solution is to only retain negative scores, so that the attacker can't build up score by using this method.

vyzo replied:
P1 is only valid while they are in the mesh, it is not retained btw.

vyzo replied:
So your attack is not plausible against P1, this simply disappears when they leave the mesh.
It is however a valid concern about building score by publishing messages with the first message deliveries counter.

raulk replied:
@vyzo 

> P1 is only valid while they are in the mesh, it is not retained btw.

👍

> The simple (and correct) solution is to only retain negative scores, so that the attacker can't build up score by using this method.

I think we need something more sophisticated, though. I wouldn't like a peer to lose all their positive karma due to a 1-sec hiccup. That's why I suggested some form of failure counting and tolerance, or live/dead time tallying, after which, yes, we'd reset any positive scores.

vyzo replied:
Following sync discussion, it's hard to build logic around disconnections because we don't know the reason of the disconnection -- it could just be the connection manager kicking in and us killing the connection.

vyzo replied:
I implemented negative score retention; take another look.
When we have a better idea how to retain positive scores, we can add the relevant logic.

raulk commented:
First batch of comments; I'll continue my review later!

raulk reviewed (commented):
First batch of comments; I'll continue my review later!

vyzo committed:
move score parameter types to their own file

vyzo committed:
make PX optional and disabled by default, gate by acceptPXThreshold

vyzo commented:
Made PX emission optional (to be enabled in bootstrappers) and added a gating threshold (>= 0) for accepting PX in order to address @lanzafame's gossiping wolves attack.

vyzo committed:
use a struct for aggregating score thresholds

vyzo committed:
reraise GossipSubPrunePeers to 16

so that bootstrappers can give us a good enough chunk of peers to connect to.

vyzo committed:
rename score debug option, address some comments around inspection

- Use nil for the never fire channel when there is no inspection
- Defer a score inspection for shutdown to get one final sample on shutdown

vyzo committed:
wrap topic param validation error

vyzo committed:
godoc nits for parameter weights.

vyzo committed:
topic score cap

vyzo committed:
only retain negative scores

vyzo committed:
increase connection manager weight for mesh peers

vyzo committed:
don't forward messages with source ourselves but not published by ourselves

per stebalien's request.

vyzo committed:
gomod: use go-multiaddr-net v0.1.3

vyzo committed:
use manet.ToIP instead of hand-rolling multiaddr to IP conversion

vyzo committed:
add IP whitelist for colocation factor penalty

vyzo committed:
use named constants for tracing message rejection reasons

vyzo committed:
rename delivery_ constants to satisfy the camelcase worshipers and nitpickers

vyzo committed:
move self-publish check to pubsub

vyzo committed:
better handling of message originating from self but forwarded by another peer

vyzo committed:
correctly attribute invalid message deliveries due to self origin rejection

vyzo committed:
update REAMDE

vyzo committed:
support direct peering agreements

vyzo commented:
Merged support for direct peering agreements from #273.

raulk commented:
Taking another look now.

raulk commented on gossipsub.go:
```diff
@@ -208,17 +371,34 @@ func (gs *GossipSubRouter) handleIHave(p peer.ID, ctl *pb.ControlMessage) []*pb.
-			msg, ok := gs.mcache.Get(mid)
-			if ok {
-				ihave[mid] = msg
+			msg, count, ok := gs.mcache.GetForPeer(mid, p)
```
This creates an entry in the mcache for each message ID the peer wants. Can't a peer just spoof random message IDs, and send us huge IWANTs up to the max message size, to make us OOM?

raulk replied:
One way to prevent this is to track the IHAVEs we've sent, and interesect the incoming IWANTs against them. If the peer is asking for random stuff, we need to decrement their score or blacklist them entirely.

vyzo replied:
It will only create an entry _iff_ we have the message in mcache, and it will last for just a few seconds (until we delete the message from the cache).
I don't think it's a plausible attack vector.

raulk committed:
comments and nits.

vyzo committed:
eagerly send gossip at the end heartbeat

instead of trying to piggyback with outgoing messages

vyzo committed:
use a configurable variable for direct peer reconnect ticks

vyzo committed:
opportunistic grafting

vyzo committed:
fix typo

Co-Authored-By: Raúl Kripalani <raul@protocol.ai>

vyzo committed:
fix typo

Co-Authored-By: Raúl Kripalani <raul@protocol.ai>

vyzo commented:
Merged #279 which implements opportunistic grafting.

vyzo committed:
configurable number of peers for opportunistic grafting

vyzo committed:
disable flaky and irrelevant tests

vyzo commented:
the tests are now green on travis (check https://travis-ci.com/github/libp2p/go-libp2p-pubsub), but somehow the integration with github seems broken and has failed to report it.

vyzo commented:
ah nm, heisenbug -- it just got reported.

daviddias reviewed (approved):
We've discussed this offline and agree that the next step is:

- Merge this PR to master to support integrations in other projects
- Hold on releasing it as a new version for now.

People on board @vyzo @raulk @Stebalien

raulk commented:
Can we open an issue to track unit tests? Testground tests are showing a lot of improvement, but we really need regression tests that will run on CI.

cc @daviddias

vyzo commented:
Agreed -- @dirkmc is working on extending the coverage with unit tests.

vyzo commented:
Opened issue #281 for tracking unit test coverage.

vyzo committed:
fix bug in peerScore.setIPs

It was removing all old addresses, not just the ones in the new addres slice

vyzo committed:
gomod: update go-multiaddr-net

vyzo committed:
fix issue with unspecific IPv6 addrs in conn IP tracking

Also, don't track loopback addrs.

vyzo committed:
more rigorously defeat unspecified address gremlins

vyzo committed:
fix IPv4 -vs- IPv6 address handling in peer address tracking

Stebalien reviewed (dismissed):
Predicated on a _plan_ for landing tests. A plan that we stick to.

vyzo commented:
@Stebalien we have extensively tested on TestGround[TM]. 
Unit tests can only do so much, we can't test emergent behaviour and the reaction to various attacks with them. But we are working on adding basic unit tests that can catch regressions.

daviddias commented:
With regards to coverage

<img width="1142" alt="image" src="https://user-images.githubusercontent.com/1211152/79319980-f7107580-7f00-11ea-9448-bc33823e38e5.png">

Seems that gossipsub is well covered above 70% and this is before @dirkmc latest PR to add more tests

It would be great to get to a minimum of 70% across the board. @vyzo is this something you can work on? It will help immensely future contributors that want to make changes without breaking the codebase.

vyzo commented:
@david -- we are working with @dirkmc to ensure good coverage for the v1.1 changes.
As for the rest, agreed that it would be nice to have. I will work on it time permitting, but might take a little while.

vyzo committed:
gomod tidy

vyzo commented:
Rebased on master for merging.

daviddias commented:
![](https://media.giphy.com/media/lMameLIF8voLu8HxWV/giphy.gif)
![](https://media.giphy.com/media/5PSPV1ucLX31u/giphy.gif)
![](https://media.giphy.com/media/AwcmOV28QPnck/giphy.gif)

YEaasasssss!!! Phenomenal work everyone!!!

bitcard commented:
mark

Stebalien commented:
@bitcard please stop commenting "mark" on random issues. If you want to bookmark something, save the link somewhere. Everyone subscribed to these issues gets an email whenever you do this.
